File: tests/test_live_sensex.py
"""
Test live SENSEX market data with specific tokens
Target contracts at 14:22:23 on Oct 17, 2025:
1. SENSEX 30 Oct Fut: 83847
2. SENSEX 83900 CE 23 Oct: 486
3. SENSEX 83800 PE 23 Oct: 340
4. SENSEX 83700 PE 23 Oct: 304
5. SENSEX 84000 CE 23 Oct: 430
"""

import sys
import os
import time
from datetime import datetime
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from connection import BSEMulticastConnection
from packet_receiver import PacketReceiver
from decoder import PacketDecoder
from decompressor import NFCASTDecompressor
from data_collector import MarketDataCollector
from saver import DataSaver
import json

# Target tokens
TARGET_TOKENS = {
    861384: {'name': 'SENSEX 30-OCT-2025 FUT', 'expected_ltp': 83847},
    878196: {'name': 'SENSEX 83900 CE 23-OCT', 'expected_ltp': 486},
    878015: {'name': 'SENSEX 83800 PE 23-OCT', 'expected_ltp': 340},
    877845: {'name': 'SENSEX 83700 PE 23-OCT', 'expected_ltp': 304},
    877761: {'name': 'SENSEX 84000 CE 23-OCT', 'expected_ltp': 430},
}

def test_live_sensex():
    """Test with live market data for specific SENSEX contracts"""
    
    print("="*100)
    print("LIVE SENSEX MARKET DATA TEST")
    print("="*100)
    print(f"Start Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print()
    print("Target Contracts:")
    for token, info in TARGET_TOKENS.items():
        print(f"  {token}: {info['name']:<30} Expected LTP: ₹{info['expected_ltp']:>8,.2f}")
    print()
    
    # Initialize components
    config_path = os.path.join(os.path.dirname(__file__), '..', 'config.json')
    with open(config_path, 'r') as f:
        config = json.load(f)
    
    # Create filtered token map with only target tokens
    token_map = {}
    full_token_path = os.path.join(os.path.dirname(__file__), '..', 'data', 'tokens', 'token_details.json')
    with open(full_token_path, 'r') as f:
        full_map = json.load(f)
        for token_str in TARGET_TOKENS.keys():
            if str(token_str) in full_map:
                token_map[str(token_str)] = full_map[str(token_str)]
    
    print(f"Loaded {len(token_map)} target tokens from token master")
    print()
    
    # Initialize pipeline
    connection = BSEMulticastConnection(
        multicast_ip=config['multicast']['ip'],
        port=config['multicast']['port'],
        buffer_size=config.get('buffer_size', 2048)
    )
    sock = connection.connect()
    receiver = PacketReceiver(sock, config, token_map)
    decoder = PacketDecoder()
    decompressor = NFCASTDecompressor()
    collector = MarketDataCollector(token_map)
    
    # Create timestamped output filename
    timestamp_str = datetime.now().strftime('%Y%m%d_%H%M%S')
    
    # Create test_results directory if it doesn't exist
    test_results_dir = os.path.join(os.path.dirname(__file__), '..', 'data', 'test_results')
    os.makedirs(test_results_dir, exist_ok=True)
    
    csv_path = os.path.join(test_results_dir, f'sensex_live_{timestamp_str}.csv')
    json_path = os.path.join(test_results_dir, f'sensex_live_{timestamp_str}.json')
    
    # Use standard DataSaver (it will create daily files)
    # We'll move/copy the files after with timestamp
    saver = DataSaver(output_dir=os.path.join(os.path.dirname(__file__), '..', 'data'))
    
    print(f"Output will be saved to:")
    print(f"  Daily CSV:  data/processed_csv/{datetime.now().strftime('%Y%m%d')}_quotes.csv")
    print(f"  Daily JSON: data/processed_json/{datetime.now().strftime('%Y%m%d')}_quotes.json")
    print(f"  (Will copy to test_results/{timestamp_str} after)")
    print()
    
    # Track found tokens
    found_tokens = {}
    packets_processed = 0
    start_time = time.time()
    max_duration = 30  # Run for 30 seconds max
    
    print("Listening for packets... (Ctrl+C to stop)")
    print("-"*100)
    
    try:
        while time.time() - start_time < max_duration:
            # Receive packet
            packet_data = receiver.receive_packet()
            if not packet_data:
                continue
            
            packets_processed += 1
            
            # Decode packet
            decoded = decoder.decode_packet(packet_data)
            if not decoded or 'records' not in decoded:
                continue
            
            # Process each record
            for record in decoded['records']:
                token = record.get('token')
                
                # Only process target tokens
                if token not in TARGET_TOKENS:
                    continue
                
                # Decompress (if needed)
                # For now, just use base LTP from decoder
                ltp_paise = record.get('ltp', 0)
                ltp_rupees = ltp_paise / 100.0
                
                # Collect market data
                quote = collector.collect_quote(record, decoded['header'])
                
                if quote:
                    # Save to CSV/JSON
                    saver.save_quote(quote)
                    
                    # Track this token
                    if token not in found_tokens:
                        found_tokens[token] = []
                    
                    found_tokens[token].append({
                        'timestamp': datetime.now(),
                        'ltp': ltp_rupees,
                        'volume': record.get('volume', 0),
                        'trades': record.get('num_trades', 0)
                    })
                    
                    # Print update
                    expected = TARGET_TOKENS[token]['expected_ltp']
                    diff_pct = abs(ltp_rupees - expected) / expected * 100 if expected > 0 else 0
                    status = "✅" if diff_pct < 5 else "⚠️" if diff_pct < 10 else "❌"
                    
                    print(f"{status} Token {token}: {TARGET_TOKENS[token]['name']:<30} "
                          f"LTP: ₹{ltp_rupees:>8,.2f} (Expected: ₹{expected:>8,.2f}, "
                          f"Diff: {diff_pct:>5.1f}%)")
            
            # Show progress every 10 packets
            if packets_processed % 10 == 0:
                print(f"  ... {packets_processed} packets processed, "
                      f"{len(found_tokens)}/{len(TARGET_TOKENS)} tokens found ...")
            
            # Stop if we found all tokens at least once
            if len(found_tokens) >= len(TARGET_TOKENS):
                updates_per_token = min(len(updates) for updates in found_tokens.values())
                if updates_per_token >= 3:  # Got at least 3 updates for each
                    print()
                    print("✅ All target tokens found with multiple updates!")
                    break
    
    except KeyboardInterrupt:
        print("\n\n⚠️  Interrupted by user")
    
    finally:
        connection.close()
    
    # Copy output files with timestamp
    import shutil
    daily_csv = os.path.join(os.path.dirname(__file__), '..', 'data', 'processed_csv', f'{datetime.now().strftime("%Y%m%d")}_quotes.csv')
    daily_json = os.path.join(os.path.dirname(__file__), '..', 'data', 'processed_json', f'{datetime.now().strftime("%Y%m%d")}_quotes.json')
    
    if os.path.exists(daily_csv):
        shutil.copy2(daily_csv, csv_path)
        print(f"\n✅ CSV copied to: {csv_path}")
    
    if os.path.exists(daily_json):
        shutil.copy2(daily_json, json_path)
        print(f"✅ JSON copied to: {json_path}")
    
    # Summary
    print()
    print("="*100)
    print("SUMMARY")
    print("="*100)
    print(f"Duration: {time.time() - start_time:.1f} seconds")
    print(f"Packets processed: {packets_processed}")
    print(f"Tokens found: {len(found_tokens)}/{len(TARGET_TOKENS)}")
    print()
    
    if found_tokens:
        print("Final Values:")
        print("-"*100)
        for token in TARGET_TOKENS.keys():
            if token in found_tokens:
                updates = found_tokens[token]
                latest = updates[-1]
                expected = TARGET_TOKENS[token]['expected_ltp']
                diff_pct = abs(latest['ltp'] - expected) / expected * 100 if expected > 0 else 0
                
                status = "✅ PASS" if diff_pct < 5 else "⚠️ WARN" if diff_pct < 10 else "❌ FAIL"
                
                print(f"{status} {TARGET_TOKENS[token]['name']:<30}")
                print(f"       Latest LTP: ₹{latest['ltp']:>8,.2f}")
                print(f"       Expected:   ₹{expected:>8,.2f}")
                print(f"       Difference: {diff_pct:>5.1f}%")
                print(f"       Updates:    {len(updates)}")
                print()
            else:
                print(f"❌ MISS {TARGET_TOKENS[token]['name']:<30} - NOT FOUND")
                print()
    else:
        print("❌ No target tokens found in captured packets!")
        print("   Possible reasons:")
        print("   - Market is closed")
        print("   - No trading activity for these contracts")
        print("   - Incorrect multicast configuration")
    
    print(f"\nOutput saved to:")
    print(f"  {csv_path}")
    print(f"  {json_path}")

if __name__ == '__main__':
    test_live_sensex()

File: bse-go-hft/tests/test_live_sensex.go
// BSE Go HFT - Live SENSEX Market Data Test
// ==========================================
//
// Tests live SENSEX market data with specific target tokens.
// Validates LTP values against expected prices.
//
// ================================================================================
// USAGE
// ================================================================================
//
// BUILD:
//   cd d:\bse\bse-go-hft
//   go build -o test_live_sensex.exe ./tests/test_live_sensex.go
//
// RUN:
//   .\test_live_sensex.exe
//   .\test_live_sensex.exe -duration 60s
//   .\test_live_sensex.exe -duration 2m -port 26002
//
// PARAMETERS:
//   -duration time   Test duration (default: 30s)
//   -port int        UDP port: 26002=FO (default: 26002)
//   -ip string       Multicast IP (default: "239.1.2.5")
//
// OUTPUT:
//   - Live display of target token updates
//   - CSV file: data/test_results/sensex_live_YYYYMMDD_HHMMSS.csv
//   - Summary report with PASS/FAIL status
//
// ================================================================================

package main

import (
	"encoding/binary"
	"encoding/csv"
	"flag"
	"fmt"
	"math"
	"net"
	"os"
	"path/filepath"
	"strings"
	"time"

	"golang.org/x/net/ipv4"
)

// ================================================================================
// CONFIGURATION
// ================================================================================

const (
	DefaultMulticastIP = "239.1.2.5"
	DefaultFOPort      = 26002
	HeaderSize         = 36
	RecordSize         = 264
	BufferSize         = 2048
)

// ================================================================================
// TARGET TOKENS - SENSEX Contracts
// ================================================================================

type TargetToken struct {
	Name        string
	ExpectedLTP float64 // Expected LTP for validation (set to 0 to skip validation)
}

// These are example target tokens - update with current contract tokens
var TargetTokens = map[uint32]TargetToken{
	// SENSEX Futures
	873830: {Name: "SENSEX DEC FUT", ExpectedLTP: 0}, // December Future
	861384: {Name: "SENSEX NOV FUT", ExpectedLTP: 0}, // November Future
	842364: {Name: "SENSEX OCT FUT", ExpectedLTP: 0}, // October Future

	// SENSEX Options (sample - update with actual tokens)
	878196: {Name: "SENSEX 83900 CE", ExpectedLTP: 0},
	878015: {Name: "SENSEX 83800 PE", ExpectedLTP: 0},
	877845: {Name: "SENSEX 83700 PE", ExpectedLTP: 0},
	877761: {Name: "SENSEX 84000 CE", ExpectedLTP: 0},
}

// ================================================================================
// DATA STRUCTURES
// ================================================================================

type TickUpdate struct {
	Timestamp time.Time
	Token     uint32
	Name      string
	LTP       float64
	Volume    int32
	SeqNum    uint32
}

type TokenStats struct {
	Updates  []TickUpdate
	FirstLTP float64
	LastLTP  float64
	HighLTP  float64
	LowLTP   float64
	TotalVol int64
}

// ================================================================================
// CSV WRITER
// ================================================================================

type TestCSVWriter struct {
	file     *os.File
	writer   *csv.Writer
	filename string
}

func NewTestCSVWriter() (*TestCSVWriter, error) {
	timestamp := time.Now().Format("20060102_150405")

	// Create test_results directory
	dir := "data/test_results"
	os.MkdirAll(dir, 0755)

	filename := filepath.Join(dir, fmt.Sprintf("sensex_live_%s.csv", timestamp))

	file, err := os.Create(filename)
	if err != nil {
		return nil, err
	}

	writer := csv.NewWriter(file)

	// Write header
	headers := []string{
		"timestamp", "token", "name", "ltp", "volume", "seq_num",
		"expected_ltp", "diff_pct", "status",
	}
	writer.Write(headers)

	return &TestCSVWriter{
		file:     file,
		writer:   writer,
		filename: filename,
	}, nil
}

func (cw *TestCSVWriter) WriteUpdate(update TickUpdate, expectedLTP float64) error {
	var diffPct float64
	var status string

	if expectedLTP > 0 {
		diffPct = math.Abs(update.LTP-expectedLTP) / expectedLTP * 100
		if diffPct < 5 {
			status = "PASS"
		} else if diffPct < 10 {
			status = "WARN"
		} else {
			status = "FAIL"
		}
	} else {
		status = "N/A"
	}

	row := []string{
		update.Timestamp.Format("2006-01-02 15:04:05.000"),
		fmt.Sprintf("%d", update.Token),
		update.Name,
		fmt.Sprintf("%.2f", update.LTP),
		fmt.Sprintf("%d", update.Volume),
		fmt.Sprintf("%d", update.SeqNum),
		fmt.Sprintf("%.2f", expectedLTP),
		fmt.Sprintf("%.2f", diffPct),
		status,
	}

	return cw.writer.Write(row)
}

func (cw *TestCSVWriter) Close() {
	cw.writer.Flush()
	cw.file.Close()
}

// ================================================================================
// MULTICAST RECEIVER
// ================================================================================

func createMulticastSocket(multicastIP string, port int) (*net.UDPConn, error) {
	addr := fmt.Sprintf("%s:%d", multicastIP, port)
	gaddr, err := net.ResolveUDPAddr("udp4", addr)
	if err != nil {
		return nil, err
	}

	conn, err := net.ListenMulticastUDP("udp4", nil, gaddr)
	if err != nil {
		return nil, err
	}

	conn.SetReadBuffer(16 * 1024 * 1024)

	pconn := ipv4.NewPacketConn(conn)
	pconn.SetControlMessage(ipv4.FlagTTL|ipv4.FlagDst, true)

	return conn, nil
}

// ================================================================================
// MAIN TEST
// ================================================================================

func testLiveSensex(port int, multicastIP string, duration time.Duration) {
	fmt.Println(strings.Repeat("=", 100))
	fmt.Println("LIVE SENSEX MARKET DATA TEST")
	fmt.Println(strings.Repeat("=", 100))
	fmt.Printf("Start Time: %s\n", time.Now().Format("2006-01-02 15:04:05"))
	fmt.Printf("Duration:   %s\n", duration)
	fmt.Printf("Port:       %d (F&O)\n", port)
	fmt.Println()

	fmt.Println("Target Contracts:")
	for token, info := range TargetTokens {
		if info.ExpectedLTP > 0 {
			fmt.Printf("  %d: %-30s Expected LTP: ₹%8.2f\n", token, info.Name, info.ExpectedLTP)
		} else {
			fmt.Printf("  %d: %-30s (No expected LTP)\n", token, info.Name)
		}
	}
	fmt.Println()

	// Create CSV writer
	csvWriter, err := NewTestCSVWriter()
	if err != nil {
		fmt.Printf("❌ Failed to create CSV file: %v\n", err)
		return
	}
	defer csvWriter.Close()
	fmt.Printf("📝 Output CSV: %s\n", csvWriter.filename)
	fmt.Println()

	// Connect to multicast
	fmt.Printf("📡 Connecting to %s:%d...\n", multicastIP, port)
	conn, err := createMulticastSocket(multicastIP, port)
	if err != nil {
		fmt.Printf("❌ Failed to connect: %v\n", err)
		return
	}
	defer conn.Close()
	fmt.Println("✅ Connected!")
	fmt.Println()

	fmt.Println("Listening for packets... (Press Ctrl+C to stop early)")
	fmt.Println(strings.Repeat("-", 100))

	// Track found tokens
	foundTokens := make(map[uint32]*TokenStats)
	packetsProcessed := 0
	startTime := time.Now()
	buffer := make([]byte, BufferSize)

	for time.Since(startTime) < duration {
		conn.SetReadDeadline(time.Now().Add(100 * time.Millisecond))

		n, _, err := conn.ReadFromUDP(buffer)
		if err != nil {
			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
				continue
			}
			continue
		}

		packetsProcessed++

		if n < HeaderSize+RecordSize {
			continue
		}

		// Validate message type
		msgType := binary.LittleEndian.Uint16(buffer[8:10])
		if msgType != 2020 && msgType != 2021 {
			continue
		}

		// Parse records
		numRecords := (n - HeaderSize) / RecordSize
		if numRecords > 8 {
			numRecords = 8
		}

		for i := 0; i < numRecords; i++ {
			offset := HeaderSize + (i * RecordSize)
			if offset+RecordSize > n {
				break
			}

			record := buffer[offset : offset+RecordSize]
			token := binary.LittleEndian.Uint32(record[0:4])

			// Only process target tokens
			targetInfo, isTarget := TargetTokens[token]
			if !isTarget {
				continue
			}

			// Decode
			ltp := float64(int32(binary.LittleEndian.Uint32(record[36:40]))) / 100.0
			volume := int32(binary.LittleEndian.Uint32(record[24:28]))
			seqNum := binary.LittleEndian.Uint32(record[44:48])

			if ltp <= 0 {
				continue
			}

			update := TickUpdate{
				Timestamp: time.Now(),
				Token:     token,
				Name:      targetInfo.Name,
				LTP:       ltp,
				Volume:    volume,
				SeqNum:    seqNum,
			}

			// Track stats
			stats, exists := foundTokens[token]
			if !exists {
				stats = &TokenStats{
					FirstLTP: ltp,
					HighLTP:  ltp,
					LowLTP:   ltp,
				}
				foundTokens[token] = stats
			}

			stats.Updates = append(stats.Updates, update)
			stats.LastLTP = ltp
			if ltp > stats.HighLTP {
				stats.HighLTP = ltp
			}
			if ltp < stats.LowLTP {
				stats.LowLTP = ltp
			}
			stats.TotalVol = int64(volume)

			// Write to CSV
			csvWriter.WriteUpdate(update, targetInfo.ExpectedLTP)

			// Print update
			var status string
			if targetInfo.ExpectedLTP > 0 {
				diffPct := math.Abs(ltp-targetInfo.ExpectedLTP) / targetInfo.ExpectedLTP * 100
				if diffPct < 5 {
					status = "✅"
				} else if diffPct < 10 {
					status = "⚠️"
				} else {
					status = "❌"
				}
				fmt.Printf("%s Token %d: %-30s LTP: ₹%8.2f (Expected: ₹%8.2f, Diff: %5.1f%%)\n",
					status, token, targetInfo.Name, ltp, targetInfo.ExpectedLTP, diffPct)
			} else {
				fmt.Printf("📊 Token %d: %-30s LTP: ₹%8.2f\n", token, targetInfo.Name, ltp)
			}
		}

		// Progress every 10 packets
		if packetsProcessed%10 == 0 {
			fmt.Printf("  ... %d packets processed, %d/%d tokens found ...\n",
				packetsProcessed, len(foundTokens), len(TargetTokens))
		}

		// Stop early if we found all tokens with multiple updates
		if len(foundTokens) >= len(TargetTokens) {
			allHaveMultiple := true
			for _, stats := range foundTokens {
				if len(stats.Updates) < 3 {
					allHaveMultiple = false
					break
				}
			}
			if allHaveMultiple {
				fmt.Println()
				fmt.Println("✅ All target tokens found with multiple updates!")
				break
			}
		}
	}

	// Summary
	fmt.Println()
	fmt.Println(strings.Repeat("=", 100))
	fmt.Println("SUMMARY")
	fmt.Println(strings.Repeat("=", 100))
	fmt.Printf("Duration: %.1f seconds\n", time.Since(startTime).Seconds())
	fmt.Printf("Packets processed: %d\n", packetsProcessed)
	fmt.Printf("Tokens found: %d/%d\n", len(foundTokens), len(TargetTokens))
	fmt.Println()

	if len(foundTokens) > 0 {
		fmt.Println("Final Values:")
		fmt.Println(strings.Repeat("-", 100))

		for token, targetInfo := range TargetTokens {
			stats, found := foundTokens[token]
			if found {
				var status string
				var diffPct float64

				if targetInfo.ExpectedLTP > 0 {
					diffPct = math.Abs(stats.LastLTP-targetInfo.ExpectedLTP) / targetInfo.ExpectedLTP * 100
					if diffPct < 5 {
						status = "✅ PASS"
					} else if diffPct < 10 {
						status = "⚠️ WARN"
					} else {
						status = "❌ FAIL"
					}
				} else {
					status = "📊 DATA"
				}

				fmt.Printf("%s %s\n", status, targetInfo.Name)
				fmt.Printf("       Latest LTP: ₹%8.2f\n", stats.LastLTP)
				if targetInfo.ExpectedLTP > 0 {
					fmt.Printf("       Expected:   ₹%8.2f\n", targetInfo.ExpectedLTP)
					fmt.Printf("       Difference: %5.1f%%\n", diffPct)
				}
				fmt.Printf("       High/Low:   ₹%.2f / ₹%.2f\n", stats.HighLTP, stats.LowLTP)
				fmt.Printf("       Updates:    %d\n", len(stats.Updates))
				fmt.Println()
			} else {
				fmt.Printf("❌ MISS %s - NOT FOUND\n\n", targetInfo.Name)
			}
		}
	} else {
		fmt.Println("❌ No target tokens found in captured packets!")
		fmt.Println("   Possible reasons:")
		fmt.Println("   - Market is closed")
		fmt.Println("   - No trading activity for these contracts")
		fmt.Println("   - Incorrect multicast configuration")
		fmt.Println("   - Token IDs may have changed (check Contract Master)")
	}

	fmt.Printf("\n📝 Output saved to: %s\n", csvWriter.filename)
}

// ================================================================================
// MAIN
// ================================================================================

func main() {
	duration := flag.Duration("duration", 30*time.Second, "Test duration (e.g., 30s, 1m, 5m)")
	port := flag.Int("port", DefaultFOPort, "UDP port (default: 26002 for F&O)")
	ip := flag.String("ip", DefaultMulticastIP, "Multicast IP address")
	flag.Parse()

	testLiveSensex(*port, *ip, *duration)
}
